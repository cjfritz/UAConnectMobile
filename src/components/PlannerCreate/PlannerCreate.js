import React, { Component } from 'react';
import { Alert } from 'react-native';
import {
  Container, Content, Button, H3,
} from 'native-base';
import { connect } from 'react-redux';
import PlannerFormComponent from '../PlannerForm/PlannerForm';
import {
  plannerCreate, plannerClear, plannerValidUpdate,
} from '../../actions/PlannerActions';
import styles from './PlannerCreate.style';

export class PlannerCreate extends Component {
  static navigationOptions = {
    title: 'Planner Create',
    headerStyle: {
      backgroundColor: 'darkred',
    },
    headerTitleStyle: {
      color: 'white',
    },
  };

  componentWillMount() {
    const { clearForm, updateValidFields } = this.props;
    const propsToSetInvalid = [
      'course',
      'description',
      'term',
      'grade',
      'units',
    ];
    clearForm();
    propsToSetInvalid.forEach(prop => {
      updateValidFields({ prop, value: '' });
    });
  }

  onCreate = () => {
    Alert.alert(
      'Create Course?',
      undefined,
      [
        { text: 'Cancel' },
        { text: 'Confirm', onPress: () => this.onConfirmCreate() },
      ]
    );
  };

  onConfirmCreate = () => {
    const {
      course,
      description,
      term,
      grade,
      units,
      coursePlannerCreate,
    } = this.props;
    coursePlannerCreate({
      course,
      description,
      term,
      grade,
      units,
    });
  };

  isValid = () => {
    const { validFields } = this.props;
    const {
      validCourse,
      validDescription,
      validTerm,
      validGrade,
      validUnits,
    } = validFields;

    if (validCourse && validDescription && validTerm && validGrade && validUnits) {
      return true;
    } else {
      return false;
    }
  };

  render() {
    return (
      <Container>
        <Content>
          <PlannerFormComponent />
          <Button disabled={ !this.isValid() } style={ styles.createButton } onPress={ () => this.onCreate() }>
            <H3 style={ styles.createText }>
              Create Course
            </H3>
          </Button>
        </Content>
      </Container>
    );
  }
}

const mapDispatchToProps = dispatch => (
  ({
    coursePlannerCreate: ({
      course,
      description,
      term,
      grade,
      units,
    }) => dispatch(plannerCreate({
      course,
      description,
      term,
      grade,
      units,
    })),
    clearForm: () => dispatch(plannerClear()),
    updateValidFields: ({ prop, value }) => dispatch(plannerValidUpdate({ prop, value })),
  })
);

const mapStateToProps = state => {
  const {
    course,
    description,
    term,
    grade,
    units,
    valid,
    validFields,
  } = state.plannerForm;
  return {
    course,
    description,
    term,
    grade,
    units,
    valid,
    validFields,
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(PlannerCreate);
