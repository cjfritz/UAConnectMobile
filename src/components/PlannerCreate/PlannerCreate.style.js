import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  createButton: {
    alignItems: 'center',
    alignSelf: 'center',
    backgroundColor: 'darkred',
    justifyContent: 'center',
    marginTop: '5%',
    width: '40%',
  },
  createText: {
    color: 'white',
  },
});
