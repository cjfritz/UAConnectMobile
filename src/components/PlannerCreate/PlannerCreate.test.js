import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { PlannerCreate } from './PlannerCreate';

const props = {
  course: '',
  description: '',
  term: '',
  grade: '',
  units: '',
  validFields: {
    validCourse: false,
    validDescription: false,
    validTerm: false,
    validGrade: false,
    validUnits: false,
  },
  coursePlannerCreate: jest.fn(),
  clearForm: jest.fn(),
  updateValidFields: jest.fn(),
};

describe('<PlannerCreate />', () => {
  it('can render with empty, invalid props', () => {
    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<PlannerCreate { ...props } />);
    expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
  });

  it('can render with invalid text for grade and units props', () => {
    const newProps = {
      course: 'testCoures',
      description: 'testDesc',
      term: 'testTerm',
      grade: 'badInput',
      units: 'badInput',
      validFields: {
        validCourse: true,
        validDescription: true,
        validTerm: true,
        validGrade: false,
        validUnits: false,
      },
    };

    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<PlannerCreate { ...props } { ...newProps } />);
    expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
  });

  it('can render with invalid values greater than max range for grade and units props', () => {
    const newProps = {
      grade: 4.1,
      units: 5,
    };

    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<PlannerCreate { ...props } { ...newProps } />);
    expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
  });

  it('can render with invalid values less than min range for grade and units props', () => {
    const newProps = {
      grade: -0.5,
      units: -1,
    };

    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<PlannerCreate { ...props } { ...newProps } />);
    expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
  });
});
