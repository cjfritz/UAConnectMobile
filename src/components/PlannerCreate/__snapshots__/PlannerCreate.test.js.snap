// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlannerCreate /> can render with empty, invalid props 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Connect(PlannerForm) />
    <Styled(Button)
      disabled={true}
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "gray",
          "justifyContent": "center",
          "marginTop": "5%",
          "width": "40%",
        }
      }
    >
      <Styled(H3)
        style={
          Object {
            "color": "white",
          }
        }
      >
        Create Course
      </Styled(H3)>
    </Styled(Button)>
  </Styled(Content)>
</Styled(Container)>
`;

exports[`<PlannerCreate /> can render with invalid text for grade and units props 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Connect(PlannerForm) />
    <Styled(Button)
      disabled={true}
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "gray",
          "justifyContent": "center",
          "marginTop": "5%",
          "width": "40%",
        }
      }
    >
      <Styled(H3)
        style={
          Object {
            "color": "white",
          }
        }
      >
        Create Course
      </Styled(H3)>
    </Styled(Button)>
  </Styled(Content)>
</Styled(Container)>
`;

exports[`<PlannerCreate /> can render with invalid values greater than max range for grade and units props 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Connect(PlannerForm) />
    <Styled(Button)
      disabled={true}
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "gray",
          "justifyContent": "center",
          "marginTop": "5%",
          "width": "40%",
        }
      }
    >
      <Styled(H3)
        style={
          Object {
            "color": "white",
          }
        }
      >
        Create Course
      </Styled(H3)>
    </Styled(Button)>
  </Styled(Content)>
</Styled(Container)>
`;

exports[`<PlannerCreate /> can render with invalid values less than min range for grade and units props 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Connect(PlannerForm) />
    <Styled(Button)
      disabled={true}
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "gray",
          "justifyContent": "center",
          "marginTop": "5%",
          "width": "40%",
        }
      }
    >
      <Styled(H3)
        style={
          Object {
            "color": "white",
          }
        }
      >
        Create Course
      </Styled(H3)>
    </Styled(Button)>
  </Styled(Content)>
</Styled(Container)>
`;
